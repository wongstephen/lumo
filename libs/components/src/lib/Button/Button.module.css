.container {
  border-radius: 0.5rem;
  border: 1px solid black;
  cursor: pointer;
  padding: 0.5rem 1rem;
}
