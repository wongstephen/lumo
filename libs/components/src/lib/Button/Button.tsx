import { Slot } from '@radix-ui/react-slot';
import React from 'react';

import styles from './Button.module.css';

type ButtonSizes = 'small' | 'medium' | 'large';

type ButtonAppearance = 'primary' | 'secondary' | 'outline' | 'transparent';

export type ButtonProps = {
  /**
   * A button can have different appearances and passed custom content.
   * - 'primary (default)': Primary button action emphasis.
   * - 'secondary': Secondary button action emphasis.
   * - 'outline': Button removing the background.
   * - 'transparent': Button with no background and no border.
   *
   * @default 'primary'
   */
  appearance?: ButtonAppearance;
  /**
   * Disables the button, preventing any interaction.
   *
   * @default false
   */
  disabled?: boolean;
  /**
   * Place icon before or after its context
   *
   * @default 'before'
   */
  iconPosition?: 'before' | 'after';
  /**
   * Button size options
   *
   * @default 'medium'
   */
  size?: ButtonSizes;
  /**
   * Slot for the button content. When true, the styling will be applied to the child element.
   *
   * @default: false
   */
  asChild?: boolean;
} & React.ComponentProps<'button'>;

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      appearance = 'primary',
      disabled,
      children,
      size = 'medium',
      iconPosition = 'before',
      asChild = false,
      ...props
    },
    ref,
  ): React.JSX.Element => {
    const Comp = asChild ? Slot : 'button';

    return (
      <Comp
        className={styles['container']}
        data-appearance={appearance}
        data-disabled={disabled}
        data-size={size}
        disabled={disabled}
        ref={ref}
        {...props}
      >
        {children}
      </Comp>
    );
  },
);
Button.displayName = 'Button';

export { Button };
