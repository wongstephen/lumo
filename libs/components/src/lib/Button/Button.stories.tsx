import { expect } from '@storybook/jest';
import type { Meta, StoryObj } from '@storybook/react';
import { within } from '@storybook/testing-library';

import { Button } from './Button';

const meta: Meta<typeof Button> = {
  component: Button,
  title: 'atoms/Button',
  tags: ['autodocs'],
};
export default meta;
type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: { children: 'Click Me!' },
};

export const Secondary: Story = {
  args: { children: 'Click Me!', variant: 'secondary' },
};

export const Disabled: Story = {
  args: { children: 'Click Me!', variant: 'secondary', disabled: true },
};
