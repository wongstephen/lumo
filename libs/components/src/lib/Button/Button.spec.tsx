import { render } from '@testing-library/react';

import Button from './Button';

describe('Button', () => {
  it('should render successfully', () => {
    const { baseElement, getByRole } = render(<Button />);
    const button = getByRole('button');

    expect(button.tagName).toBe('BUTTON');
    expect(baseElement).toBeTruthy();
    expect(baseElement).toMatchSnapshot();
  });
});
