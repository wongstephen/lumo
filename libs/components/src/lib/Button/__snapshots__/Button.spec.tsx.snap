// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should render successfully 1`] = `
<body>
  <div>
    <button
      class="_container_05d42c"
      data-appearance="primary"
      data-icon-position="before"
      data-size="medium"
    />
  </div>
</body>
`;
